src/push.rs

```diff
@@ -179,22 +146,77 @@ pub async fn push_profile(data: PushProfileData<'_>) -> Result<(), PushProfileEr
             a => return Err(PushProfileError::SignExit(a)),
         };
     }
+    Ok(())
+}
 
+pub async fn build_profile_remotely(data: &PushProfileData<'_>, derivation_name: &str) -> Result<(), PushProfileError> {
     info!(
-        "Copying profile `{}` to node `{}`",
+        "Building profile `{}` for node `{}` on remote host",
         data.deploy_data.profile_name, data.deploy_data.node_name
     );
 
-    let mut copy_command = Command::new("nix");
-    copy_command.arg("copy");
+    let store_address = format!("ssh-ng://{}@{}",
+        if data.deploy_data.profile.generic_settings.ssh_user.is_some() {
+            &data.deploy_data.profile.generic_settings.ssh_user.as_ref().unwrap()
+        } else {
+            &data.deploy_defs.ssh_user
+        },
+        data.deploy_data.node.node_settings.hostname
+    );
 
-    if data.deploy_data.merged_settings.fast_connection != Some(true) {
-        copy_command.arg("--substitute-on-destination");
-    }
+    let ssh_opts_str = data.deploy_data.merged_settings.ssh_opts.join(" ");
 
-    if !data.check_sigs {
-        copy_command.arg("--no-check-sigs");
-    }
+
+    // copy the derivation to remote host so it can be built there
+    let copy_command_status = Command::new("nix").arg("copy")
+        .arg("-s")  // fetch dependencies from substitures, not localhost
+        .arg("--to").arg(&store_address)
+        .arg("--derivation").arg(derivation_name)
+        .env("NIX_SSHOPTS", ssh_opts_str.clone())
+        .stdout(Stdio::null())
+        .status()
+        .await
+        .map_err(PushProfileError::Copy)?;
+
+    match copy_command_status.code() {
+        Some(0) => (),
+        a => return Err(PushProfileError::CopyExit(a)),
+    };
+
+    let mut build_command = Command::new("nix");
+    build_command
+        .arg("build").arg(derivation_name)
+        .arg("--eval-store") .arg("auto")
+        .arg("--store").arg(&store_address)
+        .args(data.extra_build_args)
+        .env("NIX_SSHOPTS", ssh_opts_str.clone());
+
+    debug!("build command: {:?}", build_command);
+
```

Comment:
Perhaps I'm missing something, but is profile activation performed within `build_profile_remotely`?

Reply:
Oh, okay, profile activation is done within `deploy_profile`